crates/build/src/lib.rs

```diff
@@ -689,93 +700,186 @@ fn check_dylint_requirements(_working_dir: Option<&Path>) -> Result<()> {
+) -> Result<ExportSection> {
+    let filtered_exports = exports_reader.clone().into_iter().try_fold(
+        ExportSection::new(),
+        |mut exports, entry| {
+            let entry = entry.context("Parsing of wasm export section failed")?;
+            if matches!(entry.kind, wasmparser::ExternalKind::Func)
+                && (entry.name == "call" || entry.name == "deploy")
+            {
+                exports.export(entry.name, entry.kind.into(), entry.index);
+            }
+            Ok::<_, Error>(exports)
+        },
+    )?;
+
+    Ok(filtered_exports)
 }
 
-/// Load and parse a Wasm file from disk.
-fn load_module<P: AsRef<Path>>(path: P) -> Result<Module> {
+/// Load a Wasm file from disk.
+fn load_module<P: AsRef<Path>>(path: P) -> Result<Vec<u8>> {
     let path = path.as_ref();
-    parity_wasm::deserialize_file(path).context(format!(
+    fs::read(path).context(format!(
         "Loading of wasm module at '{}' failed",
         path.display(),
     ))
 }
 
+/// Encode a Wasm section payload to an output buffer.
+fn encode_module_section(payload: Payload, module: &[u8], output: &mut Vec<u8>) {
+    if let Some((id, range)) = payload.as_section() {
+        RawSection {
+            id,
+            data: &module[range],
+        }
+        .append_to(output);
+    }
+}
+
 /// Performs required post-processing steps on the Wasm artifact.
 fn post_process_wasm(
     optimized_code: &PathBuf,
     skip_wasm_validation: bool,
     verbosity: &Verbosity,
-    max_memory_pages: u32,
+    max_memory_pages: u64,
 ) -> Result<()> {
     // Deserialize Wasm module from a file.
-    let mut module =
+    let module =
         load_module(optimized_code).context("Loading of optimized wasm failed")?;
+    let mut output = Vec::new();
+    post_process_module(
+        &module,
+        skip_wasm_validation,
+        verbosity,
+        max_memory_pages,
+        &mut output,
+    )?;
+    fs::write(optimized_code, output)?;
+    Ok(())
+}
 
-    strip_exports(&mut module);
-    ensure_maximum_memory_pages(&mut module, max_memory_pages)?;
-    strip_custom_sections(&mut module);
+/// Performs required post-processing steps on the Wasm in the buffer.
+fn post_process_module(
+    module: &[u8],
+    skip_wasm_validation: bool,
+    verbosity: &Verbosity,
+    max_memory_pages: u64,
+    output: &mut Vec<u8>,
+) -> Result<()> {
+    let mut stack = Vec::new();
+
+    for payload in Parser::new(0).parse_all(module) {
+        let payload = payload?;
+
+        // Support for nested components and modules
+        match payload {
+            Payload::Version { encoding, .. } => {
+                output.extend_from_slice(match encoding {
+                    wasmparser::Encoding::Component => &wasm_encoder::Component::HEADER,
+                    wasmparser::Encoding::Module => &wasm_encoder::Module::HEADER,
+                });
+            }
+            Payload::ModuleSection { .. } | Payload::ComponentSection { .. } => {
```

Comment:
@ascjones, I think we do not support components, and we expect only one module in WASM file.
Options are to:
- parse each of them and apply post-processing rules (some operations may not have sense here like shrinking exports)
- return error
What do you think?